File: core/flowfield/src/main/java/com/w2sv/flowfield/Particle.java
package com.w2sv.flowfield;

import static processing.core.PApplet.cos;
import static processing.core.PApplet.sin;

import processing.core.PGraphics;
import processing.core.PVector;

class Particle {
    private final PVector vel;
    private final PVector acc = new PVector();
    private final PVector previousPos = new PVector();
    private final float maxSpeed;

    PVector pos;
    private boolean skipDraw = false;

    public Particle(PVector vel, float maxSpeed, PVector startPos) {
        this.vel = vel;
        this.maxSpeed = maxSpeed;
        this.pos = startPos;
        this.previousPos.set(startPos);
    }

    /**
     * Updates position, handles edge wrapping, and applies acceleration.
     */
    public void update(float angle, int width, int height) {
        acc.set(cos(angle), sin(angle));
        vel.add(acc).limit(maxSpeed);
        pos.add(vel);
        skipDraw = wrapPositionIfOutOfBounds(width, height);
    }

    /**
     * Wraps the particle around when it leaves the screen bounds.
     * Returns true if position was wrapped (to skip drawing the line).
     */
    private boolean wrapPositionIfOutOfBounds(int width, int height) {
        boolean wrapped = false;

        if (pos.x >= width) {
            pos.x -= width;
            wrapped = true;
        } else if (pos.x < 0) {
            pos.x += width;
            wrapped = true;
        }

        if (pos.y >= height) {
            pos.y -= height;
            wrapped = true;
        } else if (pos.y < 0) {
            pos.y += height;
            wrapped = true;
        }

        return wrapped;
    }

    /**
     * Draws a trail line from the previous to the current position.
     */
    public void draw(PGraphics canvas) {
        if (!skipDraw) {
            canvas.line(pos.x, pos.y, previousPos.x, previousPos.y);
        }
        skipDraw = false;
        previousPos.set(pos);
    }
}

File: core/flowfield/src/main/kotlin/com/w2sv/flowfield/simulation/Particle.kt
package com.w2sv.flowfield.simulation

import kotlin.math.cos
import kotlin.math.sin

/**
 * Particle with position, velocity and simple Euler integration.
 * previousPos is updated after drawing so the renderer can form a line segment.
 */
internal class Particle(val pos: Vec2, private val maxVelocity: Float) {
    private val vel = Vec2()
    private val acc = Vec2()
    val previousPos = Vec2(pos.x, pos.y)
    private var skipDraw = false

    /** Updates with a force specified as an angle (radians). Returns whether the position wrapped. */
    fun update(
        angle: Float,
        width: Int,
        height: Int
    ) {
        // set acceleration to unit vector of angle
        acc.setTo(cos(angle), sin(angle))
        vel.plus(acc).limit(maxVelocity)
        pos.plus(vel)
        wrapPositionIfOutOfBounds(width, height)
    }

    private fun wrapPositionIfOutOfBounds(width: Int, height: Int) {
        if (pos.x >= width) {
            pos.x -= width
            skipDraw = true
        } else if (pos.x < 0f) {
            pos.x += width
            skipDraw = true
        }
        if (pos.y >= height) {
            pos.y -= height
            skipDraw = true
        } else if (pos.y < 0f) {
            pos.y += height
            skipDraw = true
        }
    }

    /** Mark the previous() position to current pos after drawing */
    fun afterDraw() {
        previousPos.setTo(pos)
    }

    fun shouldSkipDraw(): Boolean =
        skipDraw.also { if (it) skipDraw = false }
}
